### PR TITLE
Fix ed25519 builtin program handling

### PR DESCRIPTION
**Pull request by jstarry**
#### Problem
Master and v1.9 branches are incompatible with v1.8.

The ed25519 program was originally added as a "builtin" program which means that ed25519 instructions would be loaded by the native loader. However, on master and v1.9, ed25519 was changed to be a "precompile" program to prevent it from being invoked by on-chain programs (it always returns success so this would be confusing).

On devnet/mainnet-beta, the ed25519 program has already been activated as a builtin, and so both precompile programs, secp256k1 and ed25519, are still invokable because the `prevent_calling_precompiles_as_programs` feature wasn't back ported to v1.8 and so hasn't been activated yet. In order to transition those clusters to v1.9, the precompiles need to be handled as builtins until the `prevent_calling_precompiles_as_programs` feature is activated.

#### Summary of Changes
At a high level, this PR just ensures that ed25519 is treated the same way as secp256k1. When the `prevent_calling_precompiles_as_programs` feature is activated, both programs will be transitioned from builtins to precompiles.

- Re-add ed25519 to the list of builtin programs. Since the program is already activated on all clusters, it can be added directly to the genesis builtin list.
- Ensure that ed25519 is removed as a builtin when the `prevent_calling_precompiles_as_programs` feature is activated
- Ensure that ed25519 is added as a precompile when `prevent_calling_precompiles_as_programs` feature is activated
- Add integration tests and the ability to deactivate features in program tests

Fixes https://github.com/solana-labs/solana/issues/23180

### CODE REVIEW

**Review by jackcmay — APPROVED** [review]

File: programs/ed25519-tests/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+authors = ["Solana Maintainers <maintainers@solana.foundation>"]
+repository = "https://github.com/solana-labs/solana"
+license = "Apache-2.0"
+homepage = "https://solana.com/"
+edition = "2021"
+publish = false
+
+[dev-dependencies]
+assert_matches = "1.5.0"
+ed25519-dalek = "=1.0.1"
```
Comment (rmekdma) [inline programs/ed25519-tests/Cargo.toml:13]: Is this typo?

Comment (t-nelson) [reply programs/ed25519-tests/Cargo.toml:13]: Are you a bot?  Elaborate

File: programs/ed25519-tests/tests/process_transaction.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    let instruction = new_ed25519_instruction(&privkey, message_arr);
+
+    let transaction = Transaction::new_signed_with_payer(
+        &[instruction],
+        Some(&payer.pubkey()),
+        &[payer],
+        recent_blockhash,
+    );
+
+    assert_matches!(client.process_transaction(transaction).await, Ok(()));
```
Comment (ryoqun) [inline programs/ed25519-tests/tests/process_transaction.rs:85]: hehe, i'm not familiar with those tokio based integration tests. but it seems these are nice. :)

also, it's pretty fast compared to the dreadful local-cluster thing.. ;)

```
2842-running 3 tests
2843-test test_failure ... ok
2844-test test_success ... ok
2845:test test_success_call_builtin_program ... ok
2846-
2847-test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.30s

```

so, this test is checking ed25519 instruction still success as long as it's directly embedded into tx (i.e. not via CPI) regardless `prevent_calling_precompiles_as_programs` is activated or not.

**Review by ryoqun — COMMENTED** [review]
lgtm

**Comment by behzadnouri** [comment]
Apparently this causing panic when starting the validator:
```
thread 'main' panicked at 'assertion failed: !self.freeze_started()', runtime/src/bank.rs:5383:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/core/src/panicking.rs:107:14
   2: core::panicking::panic
             at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/core/src/panicking.rs:48:5
   3: solana_runtime::bank::Bank::add_builtin_account
   4: solana_runtime::bank::Bank::add_builtin
   5: solana_runtime::bank::Bank::finish_init
   6: solana_runtime::bank::Bank::new_from_fields
   7: solana_runtime::serde_snapshot::bank_from_streams
   8: solana_runtime::snapshot_utils::rebuild_bank_from_snapshots
   9: solana_runtime::snapshot_utils::bank_from_latest_snapshot_archives
  10: solana_ledger::bank_forks_utils::load
  11: solana_core::validator::new_banks_from_ledger
  12: solana_core::validator::Validator::new
  13: solana_validator::main
```

**Comment by t-nelson** [comment]
> Apparently this causing panic when starting a validator off master on testnet:
> 
> ```
> thread 'main' panicked at 'assertion failed: !self.freeze_started()', runtime/src/bank.rs:5383:9
> stack backtrace:
>    0: rust_begin_unwind
>              at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/std/src/panicking.rs:498:5
>    1: core::panicking::panic_fmt
>              at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/core/src/panicking.rs:107:14
>    2: core::panicking::panic
>              at /rustc/6db0a0e9a4a2f55b1a85954e114ada0b45c32e45/library/core/src/panicking.rs:48:5
>    3: solana_runtime::bank::Bank::add_builtin_account
>    4: solana_runtime::bank::Bank::add_builtin
>    5: solana_runtime::bank::Bank::finish_init
>    6: solana_runtime::bank::Bank::new_from_fields
>    7: solana_runtime::serde_snapshot::bank_from_streams
>    8: solana_runtime::snapshot_utils::rebuild_bank_from_snapshots
>    9: solana_runtime::snapshot_utils::bank_from_latest_snapshot_archives
>   10: solana_ledger::bank_forks_utils::load
>   11: solana_core::validator::new_banks_from_ledger
>   12: solana_core::validator::Validator::new
>   13: solana_validator::main
> ```

Yep!  Fix is here https://github.com/solana-labs/solana/pull/23233

**Pull request merged**